astronomy/date_body.hpp

```diff
@@ -0,0 +1,781 @@
+  return day_ == month_length(year_, month_)
+             ? month_ == 12 ? Date(year_ + 1, 1, 1) : Date(year_, month_ + 1, 1)
+             : Date(year_, month_, day_ + 1);
+}
+
+constexpr Date::Date(int const year,
+                     std::int8_t const month,
+                     std::int8_t const day)
+      : year_(year),
+        month_(month),
+        day_(day) {}
+
+// Implementation of class TimeOfDay.
+
+constexpr TimeOfDay TimeOfDay::hhmmss_ns(int const hhmmss, int ns) {
+  return TimeOfDay(digit_range(hhmmss, 4, 2),
+                   digit_range(hhmmss, 2, 2),
+                   digit_range(hhmmss, 0, 2),
+                   ns).checked();
+}
+
+constexpr int TimeOfDay::hour() const {
+  return hour_;
+}
+
+constexpr int TimeOfDay::minute() const {
+  return minute_;
+}
+
+constexpr int TimeOfDay::second() const {
+  return second_;
+}
+
+constexpr int TimeOfDay::nanosecond() const {
+  return nanosecond_;
+}
+
+constexpr bool TimeOfDay::is_leap_second() const {
+  return second_ == 60;
+}
+
+constexpr bool TimeOfDay::is_end_of_day() const {
+  return hour_ == 24;
+}
+
+constexpr TimeOfDay::TimeOfDay(int const hour,
+                               int const minute,
+                               int const second,
+                               int const nanosecond)
+    : hour_(hour),
+      minute_(minute),
+      second_(second),
+      nanosecond_(nanosecond) {}
+
+constexpr TimeOfDay const& TimeOfDay::checked() const {
+  return CHECKING(
+      (hour_ == 24 && minute_ == 0 && second_ == 0 && nanosecond_ == 0) ||
+          ((nanosecond_ >= 0 && nanosecond_ <= 999'999'999) &&
+           ((hour_ == 23 && minute_ == 59 && second_ == 60) ||
+            (hour_ >= 0 && hour_ <= 23 && minute_ >= 0 && minute_ <= 59 &&
+             second_ >= 0 && second_ <= 59))),
+      *this);
+}
+
+// Implementation of class DateTime.
+
+constexpr Date const& DateTime::date() const {
+  return date_;
+}
+
+constexpr TimeOfDay const& DateTime::time() const {
+  return time_;
+}
+
+constexpr DateTime DateTime::normalized_end_of_day() const {
+  return time_.is_end_of_day()
+             ? DateTime(date_.next_day(), TimeOfDay::hhmmss_ns(00'00'00, 0))
+             : *this;
+}
+
+constexpr DateTime::DateTime(Date const date, TimeOfDay const time)
+    : date_(date),
+      time_(time) {}
+
+constexpr DateTime const& DateTime::checked() const {
+  return CHECKING(
+      !time_.is_leap_second() ||
+          (date_.day() == month_length(date_.year(), date_.month()) &&
+           (date_.month() == 6 || date_.month() == 12 || date_.month() == 3 ||
+            date_.month() == 9)),
+      *this);
+}
+
+// Date parsing.
+
+struct DateStringInfo {
+  constexpr DateStringInfo Fill() const {
+    return read == size
+               ? *this
+               : string[read] == '-'
+                     ? CHECKING(
+                           hyphens < 2,
+                           hyphens == 0
+                               ? DateStringInfo({string,
```

Comment:
We're in a macro, the commas would pester us (though if this becomes a factory call the point is moot).